[quant][core][gpu][feature] Implemented quantized conv1d cudnn op

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77175

Summary:
Previously, only quantized conv2d cudnn op has been implemented. This PR
implements the 1d variant. Because cuDNN does not have direct support
for conv1d, we have cast the 1d case to a 2d case by adding a dummy
weight dimension of 1 for the input and weight tensors. This is
analogous to how it was done for quantized cpu conv1d (e.g., see
`quantized/cpu/qconv.cpp`)

A corresponding test case was added in `test_quantized_op.py`. This
function should ideally be merged with `test_qconv1d` when cuDNN flags are
enabled and available in pytorch.

Test Plan:
```
python test/test_quantization.py -k test_qconv1d_cudnn
```

Differential Revision: [D36285804](https://our.internmc.facebook.com/intern/diff/D36285804)

dzdang committed:
[quant][core][gpu][feature] Implemented quantized conv1d cudnn op

Summary:
Previously, only quantized conv2d cudnn op has been implemented. This PR
implements the 1d variant. Because cuDNN does not have direct support
for conv1d, we have cast the 1d case to a 2d case by adding a dummy
weight dimension of 1 for the input and weight tensors. This is
analogous to how it was done for quantized cpu conv1d.

A corresponding test case was added in `test_quantized_op.py`. This
function should ideally be merged with `test_qconv1d` when cuDNN are
enabled by default.

Test Plan:
```
python test/test_quantization.py -k test_qconv1d_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36285804).

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized conv1d cudnn op"


Summary:
Previously, only quantized conv2d cudnn op has been implemented. This PR
implements the 1d variant. Because cuDNN does not have direct support
for conv1d, we have cast the 1d case to a 2d case by adding a dummy
weight dimension of 1 for the input and weight tensors. This is
analogous to how it was done for quantized cpu conv1d (e.g., see
`quantized/cpu/qconv.cpp`)

A corresponding test case was added in `test_quantized_op.py`. This
function should ideally be merged with `test_qconv1d` when cuDNN flags are
enabled and available in pytorch.

Test Plan:
```
python test/test_quantization.py -k test_qconv1d_cudnn
```

Differential Revision: [D36285804](https://our.internmc.facebook.com/intern/diff/D36285804)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized conv1d cudnn op"


Summary:
Previously, only quantized conv2d cudnn op has been implemented. This PR
implements the 1d variant. Because cuDNN does not have direct support
for conv1d, we have cast the 1d case to a 2d case by adding a dummy
weight dimension of 1 for the input and weight tensors. This is
analogous to how it was done for quantized cpu conv1d (e.g., see
`quantized/cpu/qconv.cpp`)

A corresponding test case was added in `test_quantized_op.py`. This
function should ideally be merged with `test_qconv1d` when cuDNN flags are
enabled and available in pytorch.

Test Plan:
```
python test/test_quantization.py -k test_qconv1d_cudnn
```

Differential Revision: [D36285804](https://our.internmc.facebook.com/intern/diff/D36285804)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -406,6 +429,10 @@ class QConvInt8 final {
 };
 
 TORCH_LIBRARY_IMPL(quantized, QuantizedCUDA, m) {
+  m.impl(TORCH_SELECTIVE_NAME("quantized::conv1d"), QConv1dInt8<false>::run);
+  m.impl(TORCH_SELECTIVE_NAME("quantized::conv1d_relu"), QConv1dInt8<true>::run);
```
maybe only register `quantized::conv1d.new` and `quantized::conv1d_relu.new`? I think we want to deprecate the rest

dzdang replied:
I remember doing this and it seems it was causing the issues with the dispatcher not being able to find the QuantizedCuda backend for conv1d.

dzdang replied:
ah I see. quantized cpu doesn't have .new. so we should just keep `quantized::conv1d` and get rid of `quantized::conv1d.new`

jerryzh168 replied:
`quantized::conv1d.new` is the new one, I think we should get rid of `quantized::conv1d` and just keep `quantized::conv1d.new`

dzdang replied:
yes, but what I'm saying is currnetly QuantizedCPU doesn't use `quantized::conv1d.new`. it only defines `quantized::conv1d` and it uses the packed weight signature (see https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/library.cpp#L64-L65). it's not consistent with conv2d. 

so maybe we can just change  `quantized::conv1d` in QuantizedCPU to `quantized::conv1d.new` so we're consistent with conv2d?

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -5146,6 +5148,87 @@ def test_qconv1d(
+           input_channels_per_group=st.integers(1, 32),
+           # cudnn only supports multiples of 4, but we have explicitly added padding on the backend
+           output_channels_per_group=st.integers(1, 32),
+           groups=st.integers(1, 1), # currently padding only supports groups=1
```
+1 for lint

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized conv1d cudnn op"


Summary:
Previously, only quantized conv2d cudnn op has been implemented. This PR
implements the 1d variant. Because cuDNN does not have direct support
for conv1d, we have cast the 1d case to a 2d case by adding a dummy
weight dimension of 1 for the input and weight tensors. This is
analogous to how it was done for quantized cpu conv1d (e.g., see
`quantized/cpu/qconv.cpp`)

A corresponding test case was added in `test_quantized_op.py`. This
function should ideally be merged with `test_qconv1d` when cuDNN flags are
enabled and available in pytorch.

Test Plan:
```
python test/test_quantization.py -k test_qconv1d_cudnn
```

Differential Revision: [D36285804](https://our.internmc.facebook.com/intern/diff/D36285804)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36285804).

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized conv1d cudnn op"


Summary:
Previously, only quantized conv2d cudnn op has been implemented. This PR
implements the 1d variant. Because cuDNN does not have direct support
for conv1d, we have cast the 1d case to a 2d case by adding a dummy
weight dimension of 1 for the input and weight tensors. This is
analogous to how it was done for quantized cpu conv1d (e.g., see
`quantized/cpu/qconv.cpp`)

A corresponding test case was added in `test_quantized_op.py`. This
function should ideally be merged with `test_qconv1d` when cuDNN flags are
enabled and available in pytorch.

Test Plan:
```
python test/test_quantization.py -k test_qconv1d_cudnn
```

Differential Revision: [D36285804](https://our.internmc.facebook.com/intern/diff/D36285804)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36285804).

dzdang commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -1421,8 +1421,8 @@ class QConvInt8ForBC final {
-  m.impl(TORCH_SELECTIVE_NAME("quantized::conv1d"),          QConv1dInt8<false>::run);
-  m.impl(TORCH_SELECTIVE_NAME("quantized::conv1d_relu"),     QConv1dInt8<true>::run);
+  m.impl(TORCH_SELECTIVE_NAME("quantized::conv1d.new"),          QConv1dInt8<false>::run);
+  m.impl(TORCH_SELECTIVE_NAME("quantized::conv1d_relu.new"),     QConv1dInt8<true>::run);
```
@jerryzh168 I changed the cpu one to be `.new` to be consistent with conv2d. any objections?

jerryzh168 replied:
oh, I see, so cpu does not have this `new` variant before?

jerryzh168 replied:
in that case let's just use the old one and add a comment in GPU registrations, we don't want to break bc

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized conv1d cudnn op"


Summary:
Previously, only quantized conv2d cudnn op has been implemented. This PR
implements the 1d variant. Because cuDNN does not have direct support
for conv1d, we have cast the 1d case to a 2d case by adding a dummy
weight dimension of 1 for the input and weight tensors. This is
analogous to how it was done for quantized cpu conv1d (e.g., see
`quantized/cpu/qconv.cpp`)

A corresponding test case was added in `test_quantized_op.py`. This
function should ideally be merged with `test_qconv1d` when cuDNN flags are
enabled and available in pytorch.

Test Plan:
```
python test/test_quantization.py -k test_qconv1d_cudnn
```

Differential Revision: [D36285804](https://our.internmc.facebook.com/intern/diff/D36285804)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36285804).

dzdang committed:
Update on "[quant][core][gpu][feature] Implemented quantized conv1d cudnn op"


Summary:
Previously, only quantized conv2d cudnn op has been implemented. This PR
implements the 1d variant. Because cuDNN does not have direct support
for conv1d, we have cast the 1d case to a 2d case by adding a dummy
weight dimension of 1 for the input and weight tensors. This is
analogous to how it was done for quantized cpu conv1d (e.g., see
`quantized/cpu/qconv.cpp`)

A corresponding test case was added in `test_quantized_op.py`. This
function should ideally be merged with `test_qconv1d` when cuDNN flags are
enabled and available in pytorch.

Test Plan:
```
python test/test_quantization.py -k test_qconv1d_cudnn
```

Differential Revision: [D36285804](https://our.internmc.facebook.com/intern/diff/D36285804)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36285804).
